src/psyclone/psyGen.py

```diff
@@ -4178,6 +4304,142 @@ def _number_handler(self, node, parent):
+    def kind(self):
+        return self._kind
+
+    def __str__(self):
+        return (self.name + "<" + self.datatype + "," + str(self.dimensions) +
+                "," + self.kind + ">")
+
+
+class SymbolTable(object):
+    '''
+    Encapsulates the symbols table and provides methods to declare new symbols
+    and look up existing symbols. It is implemented as a monolithic scope
+    symbol table.
+    '''
+    def __init__(self):
+        self._symbols = {}
+
+    def declare(self, name, datatype, dimensions, kind):
+        '''
+        Declare a new symbol in the symbols table.
+
+        :param name: Name of the symbol.
+        :type name: string
+        :param datatype: Data type of the symbol.
+        :type datatype: string
+        :param dimensions: Dimensions of the symbol (0 represents an scalar
+                           symbol).
+        :type dimensions: list of integers
+        :param kind: Information about the variable declaration attributes.
+        :type kind: string
+        '''
+        if name in self._symbols:
+            raise InternalError("Multiple definition of symbol: '{0}'."
+                                "".format(name))
+        else:
+            self._symbols[name] = Symbol(name, datatype, dimensions, kind)
+
+    def lookup(self, name):
+        '''
+        Look up a symbol in the symbols table.
+
+        :param name: Name of the symbol
+        :type name: string
+        :raises KeyError: If the given name is not in the Symbol Table.
+        '''
+        return self._symbols[name]
+
+    def view(self):
+        '''
+        Print a representation of this Symbol Table to stdout.
+        '''
+        print("Symbol Table:")
+        for symbol in self._symbols.values():
+            print(str(symbol))
+
+    def __str__(self):
+        return ("Symbol Table:\n" +
+                "\n".join(map(str, self._symbols.values())) +
+                "\n")
+
+
+class KernelSchedule(Schedule):
+    '''
+    A kernelSchedule inherits the functionality from Schedule and adds a symbol
+    table to keep a record of the declared variables and their attributes.
+
+    :param name: Kernel subroutine name
+    :type name: string
+    '''
+
+    def __init__(self, name):
+        super(KernelSchedule, self).__init__(None, None)
+        self._name = name
+        self._symbol_table = SymbolTable()
+
+    @property
+    def symbol_table(self):
+        return self._symbol_table
+
+    def view(self, indent=0):
+        '''
+        Print a text representation of this node to stdout and then
+        call the view() method of any children.
+
+        :param indent: Depth of indent for output text
+        :type indent: integer
+        '''
+        print(self.indent(indent) + self.coloured_text + "[name:'" + self._name
+              + "']")
+        for entity in self._children:
+            entity.view(indent=indent + 1)
+
+    def __str__(self):
+        result = "Schedule[name:'" + self._name + "']:\n"
```

Comment:
I'm not sure whether we should be outputting "KernelSchedule" instead of "Schedule". Maybe by the end of this review I'll have decided :-)

Reply:
I've decided I'm happy that it's still a Schedule.